glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/PingUploadWorker.kt

```diff
@@ -107,12 +113,19 @@ class PingUploadWorker(context: Context, params: WorkerParameters) : Worker(cont
                     // If the status is `null` there was some kind of unrecoverable error
                     // so we return a known unrecoverable error status code
                     // which will ensure this gets treated as such.
-                    val status = Glean.httpClient.doUpload(
+                    val result = Glean.httpClient.doUpload(
                             action.request.path,
                             action.request.body,
                             action.request.headers,
                             Glean.configuration
-                    ) ?: UNRECOVERABLE_ERROR_STATUS_CODE
+                    )
+
+                    // Translate the upload result into a status code the Rust side understands.
+                    val status = when (result) {
+                        is RecoverableFailure -> RECOVERABLE_ERROR_STATUS_CODE
```

Comment:
I'd really like this part to live in Rust, rather than here (we can still do this as a follow-up). The logic in Rust will need to make a decision, in the future, in order to potentially record statistics about this. I'd be happy for us to just file a bug for it, no need to do it right now.

Reply:
I want to move it over as well, but I didn't want to blow up this PR even further.

Reply:
done: https://bugzilla.mozilla.org/show_bug.cgi?id=1628700